File: oss/clawpipe/swift-sdk/Sources/ClawPipe/Gateway.swift
/// Gateway — URLSession HTTP client for the ClawPipe gateway API.
///
/// POSTs to https://api.clawpipe.ai/v1/prompt and decodes the JSON response.

import Foundation

// MARK: - Error type

public struct GatewayError: Error, LocalizedError {
    public let statusCode: Int
    public let body: String

    public var errorDescription: String? {
        "ClawPipe gateway error \(statusCode)\(body.isEmpty ? "" : ": \(body.prefix(200))")"
    }
}

// MARK: - Gateway

public struct Gateway: Sendable {

    private let baseURL: URL
    private let apiKey: String
    private let projectId: String
    private let session: URLSession

    /// - Parameters:
    ///   - baseURL: Root URL of the gateway (defaults to `https://api.clawpipe.ai/v1`).
    ///   - apiKey:  ClawPipe project API key.
    ///   - projectId: ClawPipe project identifier.
    ///   - session: URLSession to use (injectable for testing).
    public init(
        baseURL: URL = URL(string: "https://api.clawpipe.ai/v1")!,
        apiKey: String,
        projectId: String,
        session: URLSession = .shared
    ) {
        self.baseURL = baseURL
        self.apiKey = apiKey
        self.projectId = projectId
        self.session = session
    }

    /// Dispatch a prompt to the gateway and return a `PipelineResult`.
    public func call(
        prompt: String,
        provider: String,
        model: String
    ) async throws -> PipelineResult {
        let requestBody = GatewayRequest(
            prompt: prompt,
            provider: provider,
            model: model,
            projectId: projectId
        )
        let url = baseURL.appendingPathComponent("prompt")
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        urlRequest.setValue(projectId, forHTTPHeaderField: "X-Project-Id")
        urlRequest.httpBody = try JSONEncoder().encode(requestBody)

        let start = Date()
        let (data, response) = try await session.data(for: urlRequest)
        let latency = Date().timeIntervalSince(start) * 1_000

        guard let http = response as? HTTPURLResponse else {
            throw GatewayError(statusCode: -1, body: "Non-HTTP response")
        }
        guard http.statusCode < 400 else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw GatewayError(statusCode: http.statusCode, body: body)
        }

        let decoded = try JSONDecoder().decode(GatewayResponse.self, from: data)
        return PipelineResult(
            text: decoded.text,
            tokensIn: decoded.tokensIn,
            tokensOut: decoded.tokensOut,
            latencyMs: latency,
            boosted: false,
            cached: decoded.cached
        )
    }
}

File: oss/clawpipe/dotnet-sdk/ClawPipe/Gateway.cs
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ClawPipe;

/// <summary>
/// HTTP client for the ClawPipe gateway API.
/// Posts prompts to https://api.clawpipe.ai/v1/prompt and returns structured results.
/// </summary>
public sealed class Gateway : IDisposable
{
    private readonly HttpClient _http;
    private readonly string _apiKey;
    private readonly string _gatewayUrl;

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    public Gateway(string apiKey, string? gatewayUrl = null, HttpClient? httpClient = null)
    {
        _apiKey = apiKey;
        _gatewayUrl = (gatewayUrl ?? "https://api.clawpipe.ai/v1").TrimEnd('/');
        _http = httpClient ?? new HttpClient { Timeout = TimeSpan.FromSeconds(60) };
    }

    /// <summary>Send a prompt to the gateway and return a PipelineResult.</summary>
    public async Task<PipelineResult> CallAsync(
        string prompt,
        string provider = "openai",
        string model = "gpt-4o-mini",
        CancellationToken ct = default)
    {
        var payload = new GatewayRequest(prompt, provider, model);
        var json = JsonSerializer.Serialize(payload, JsonOpts);
        using var req = new HttpRequestMessage(HttpMethod.Post, $"{_gatewayUrl}/prompt")
        {
            Content = new StringContent(json, Encoding.UTF8, "application/json"),
        };
        req.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", _apiKey);

        var sw = System.Diagnostics.Stopwatch.StartNew();
        using var resp = await _http.SendAsync(req, ct).ConfigureAwait(false);
        sw.Stop();

        if (!resp.IsSuccessStatusCode)
        {
            var body = await resp.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
            throw new GatewayException((int)resp.StatusCode, body);
        }

        var envelope = await resp.Content.ReadFromJsonAsync<GatewayResponse>(JsonOpts, ct).ConfigureAwait(false)
            ?? throw new InvalidOperationException("Empty response from gateway");

        return new PipelineResult(
            Text: envelope.Text,
            TokensIn: envelope.TokensIn,
            TokensOut: envelope.TokensOut,
            LatencyMs: sw.Elapsed.TotalMilliseconds,
            Boosted: false,
            Cached: false
        );
    }

    public void Dispose() => _http.Dispose();

    // ── internal DTO ──────────────────────────────────────────────────────

    private sealed record GatewayRequest(string Prompt, string Provider, string Model);
}
